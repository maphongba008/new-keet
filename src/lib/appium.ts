import { ViewProps } from 'react-native'
import _includes from 'lodash/includes'

type NamedEnum<Key extends string> = { [v in Key]: v }

function inferGeneric<T extends string>(t: NamedEnum<T>) {
  return t
}

export const APPIUM_IDs = inferGeneric({
  back_button: 'back_button',
  // LobbyScreen
  lobby_btn_join_room: 'lobby_btn_join_room',
  lobby_btn_create_room: 'lobby_btn_create_room',
  lobby_btn_create_chat: 'lobby_btn_create_chat',
  lobby_btn_create_broadcast: 'lobby_btn_create_broadcast',
  lobby_btn_create_back: 'lobby_btn_create_back',
  lobby_input_room_name: 'lobby_input_room_name',
  lobby_input_room_description: 'lobby_input_room_description',
  lobby_btn_paste: 'lobby_btn_paste',
  lobby_btn_qr: 'lobby_btn_qr',
  lobby_btn_submit: 'lobby_btn_submit',
  lobby_btn_add_room: 'lobby_btn_add_room',
  lobby_tap_to_call: 'lobby_tap_to_call',
  lobby_input_search: 'lobby_input_search',
  lobby_btn_clear_search: 'lobby_btn_clear_search',
  lobby_btn_rooms: 'lobby_btn_rooms',
  lobby_btn_profile: 'lobby_btn_profile',
  lobby_dm: 'lobby_dm',
  lobby_filter: 'lobby_filter',
  lobby_selected_filter: 'lobby_selected_filter',
  lobby_joining_rooms: 'lobby_joining_rooms',
  lobby_swipable_btn_mute: 'lobby_swipable_btn_mute',
  lobby_swipable_btn_leave: 'lobby_swipable_btn_leave',
  lobby_bottomsheet_btn_leave: 'lobby_bottomsheet_btn_leave',
  // CreateBroadcastRoomScreen
  broadcast_accept_admin: 'broadcast_accept_admin',
  broadcast_btn_create: 'broadcast_btn_create',
  // RoomScreen
  room_call_join: 'room_call_join',
  room_call_mute: 'room_call_mute',
  room_call_video: 'room_call_video',
  room_call_speaker: 'room_call_speaker',
  room_call_end: 'room_call_end',
  room_btn_attach: 'room_btn_attach',
  room_btn_send: 'room_btn_send',
  room_input_msg: 'room_input_msg',
  room_attach_doc: 'room_attach_doc',
  room_attach_media: 'room_attach_media',
  room_capture_video: 'room_capture_video',
  room_capture_img: 'room_capture_img',
  room_btn_edit: 'room_btn_edit',
  room_btn_reply: 'room_btn_reply',
  room_btn_download: 'room_btn_download',
  room_btn_copy_msg: 'room_btn_copy_msg',
  room_btn_copy_msg_link: 'room_btn_copy_msg_link',
  room_btn_copy_img: 'room_btn_copy_img',
  room_btn_share: 'room_btn_share',
  room_btn_forward: 'room_btn_forward',
  room_btn_flag: 'room_btn_flag',
  room_btn_del: 'room_btn_del',
  room_btn_clear: 'room_btn_clear',
  room_btn_invite: 'room_btn_invite',
  room_btn_pin: 'room_btn_pin',
  room_profile_name: 'room_profile_name',
  room_profile_avatar: 'room_profile_avatar',
  room_pinned_message_header: 'room_pinned_message_header',
  // RoomOptionsScreen
  options_edit_details: 'options_edit_details',
  options_btn_files: 'options_btn_files',
  options_btn_share_link: 'options_btn_share_link',
  options_btn_share_qr: 'options_btn_share_qr',
  options_btn_clear_files: 'options_btn_clear_files',
  options_btn_leave: 'options_btn_leave',
  options_btn_toggle_notifications: 'options_btn_toggle_notifications',
  options_btn_toggle_call: 'options_btn_toggle_call',
  // InviteQRScreen
  invite_btn_type: 'invite_btn_type',
  invite_btn_expires_hours: 'invite_btn_expires_hours',
  invite_btn_expires_days: 'invite_btn_expires_days',
  invite_btn_expires_weeks: 'invite_btn_expires_weeks',
  invite_btn_no_expire: 'invite_btn_no_expire',
  invite_btn_accept_admin: 'invite_btn_accept_admin',
  invite_btn_generate: 'invite_btn_generate',
  // RoomMembersScreen
  members_search: 'members_search',
  // AccountScreen
  profile_pick_avatar: 'profile_pick_avatar',
  profile_edit_name: 'profile_edit_name',
  profile_network_status: 'profile_network_status',
  account_btn_share_network_info: 'account_btn_share_network_info',
  account_btn_copy_network_info: 'account_btn_copy_network_info',
  profile_tos: 'profile_tos',
  // list
  list_item: 'list_item_' as any,
  create_channel_btn: 'create_channel_btn',
  // Avatar
  avatar_image: 'avatar_image',
  // onboarding
  onboarding_btn_skip: 'onboarding_btn_skip',
  onboarding_btn_next: 'onboarding_btn_next',
  onboarding_btn_prev: 'onboarding_btn_prev',
  onboarding_btn_finish: 'onboarding_btn_finish',
  onboarding_tos_btn_let_in: 'onboarding_tos_btn_let_in',
  onboarding_input_username: 'onboarding_input_username',
  onboarding_btn_save_username: 'onboarding_btn_save_username',
  // IdentityScreen
  identity_btn_create: 'identity_btn_create',
  identity_btn_link: 'identity_btn_link',
  identity_btn_skip: 'identity_btn_skip',
  // Room Participant list
  participant_item: 'participant_item_' as any,
  participant_btn_all: 'participant_btn_all',
  // Tooltips
  tooltip_btn_profile: 'tooltip_btn_profile',
  tooltip_btn_join_room: 'tooltip_btn_join_room',
  // PasscodeScreen
  passcode_btn_back: 'passcode_btn_back',
  passcode_btn_close: 'passcode_btn_close',
  passcode_btn_delete: 'passcode_btn_delete',
  passcode_confirm_btn_back: 'passcode_confirm_btn_back',
  passcode_confirm_btn_close: 'passcode_confirm_btn_close',
  // RecoverAccountScreen
  recover_btn_link: 'recover_btn_link',
  recover_btn_recover: 'recover_btn_recover',
  recover_input_passphrase: 'recover_input_passphrase',
  recover_btn_continue: 'recover_btn_continue',
  recover_btn_paste: 'recover_btn_paste',
  recover_btn_success: 'recover_btn_success',
  // IdentitySyncScreen
  identity_accept_checkbox: 'identity_accept_checkbox',
  identity_btn_accept: 'identity_btn_accept',
  identity_btn_use_camera_action: 'identity_btn_use_camera_action',
  identity_btn_use_invite_link: 'identity_btn_use_invite_link',
  identity_sync_invite_link_input: 'identity_sync_invite_link_input',
  identity_sync_invite_link_submit: 'identity_sync_invite_link_submit',
  identity_sync_invite_link_declined: 'identity_sync_invite_link_declined',
  identity_sync_invite_request_confirm: 'identity_sync_invite_request_confirm',
  identity_sync_invite_request_notMyDevice:
    'identity_sync_invite_request_notMyDevice',
  identity_sync_invite_request_success: 'identity_sync_invite_request_success',
  // Create ID
  create_identity_agreement_btn: 'create_identity_agreement_btn',
  create_identity_view_phrase_btn: 'create_identity_view_phrase_btn',
  create_identity_copy_phrase_btn: 'create_identity_copy_phrase_btn',
  create_identity_setup_continue_btn: 'create_identity_setup_continue_btn',
  create_id_verification_word: 'create_id_verification_word',
  create_id_submit_verification: 'create_id_submit_verification',
  create_id_onFinish_btn: 'create_id_onFinish_btn',
  // Components
  radio_btn: 'radio_btn',
  radio_btn_checked: 'radio_btn_checked',
  // Call Screen
  call_shared_screen_expand_button: 'call_shared_screen_expand_button',
  // MediaPreviewScreen
  media_preview_video_view: 'media_preview_video_view',
  media_preview_btn_custom_play: 'media_preview_btn_custom_play',
  media_preview_loading: 'media_preview_loading',
  // Discover Community
  discover_community_join_room: 'discover_community_join_room',
  discover_community_select_room: 'discover_community_select_room',

  // wallet
  wallet_default_server_action: 'wallet_default_server_action',
  wallet_custom_server_action: 'wallet_custom_server_action',
  wallet_server_get_started: 'wallet_server_get_started',
  wallet_setup_onFinish_btn: 'wallet_setup_onFinish_btn',
  wallet_create_agreement_btn: 'wallet_create_agreement_btn',
  wallet_settings_option_any: 'wallet_settings_option_any',
  wallet_settings_option_custom: 'wallet_settings_option_custom',
  wallet_settings_amount_input: 'wallet_settings_amount_input',
  wallet_settings_save_btn: 'wallet_settings_save_btn',
  wallet_send_submit_btn: 'wallet_send_submit_btn',
  wallet_send_scanner_btn: 'wallet_send_scanner_btn',
  wallet_send_scanner_back_btn: 'wallet_send_scanner_back_btn',
  wallet_send_open_scanner_btn: 'wallet_send_open_scanner_btn',
  wallet_send_paste_btn: 'wallet_send_paste_btn',
  wallet_input_wallet_address: 'wallet_input_wallet_address',
  wallet_input_transaction_description: 'wallet_input_transaction_description',
  wallet_btn_transfer_next: 'wallet_btn_transfer_next',
  wallet_btn_select_keet_contact: 'wallet_btn_select_keet_contact',
  wallet_btn_select_crypto_wallet: 'wallet_btn_select_crypto_wallet',
  wallet_receive_generate_invoice_btn: 'wallet_receive_generate_invoice_btn',
  wallet_receive_copy_invoice_btn: 'wallet_receive_copy_invoice_btn',
  wallet_receive_share_invoice_btn: 'wallet_receive_share_invoice_btn',
  wallet_custom_server_set_mode_btn: 'wallet_custom_server_set_mode_btn',
  wallet_custom_server_continue_btn: 'wallet_custom_server_continue_btn',
  wallet_custom_server_host_input: 'wallet_custom_server_host_input',
  wallet_custom_server_port_input: 'wallet_custom_server_port_input',
  wallet_custom_server_eth_indexer_input:
    'wallet_custom_server_eth_indexer_input',
  wallet_custom_server_connect_btn: 'wallet_custom_server_connect_btn',
  // User Profile Screen
  user_profile_chat: 'user_profile_chat',
  user_profile_call: 'user_profile_call',
  user_profile_send_dm: 'user_profile_send_dm',
} as const)

const DYNAMIC_KEY = [APPIUM_IDs.list_item]

export const appiumTestProps = (
  id?: keyof typeof APPIUM_IDs | string,
): Pick<ViewProps, 'accessible' | 'testID' | 'accessibilityLabel'> => {
  try {
    if (!id) {
      return {}
    }

    const appId = APPIUM_IDs[id as keyof typeof APPIUM_IDs]
    if (appId)
      return { accessible: true, testID: appId, accessibilityLabel: appId }

    if (DYNAMIC_KEY.some((substring) => _includes(id, substring))) {
      return { accessible: true, testID: id, accessibilityLabel: id }
    }
    return {}
  } catch (e) {
    console.log(e)
    return {}
  }
}
