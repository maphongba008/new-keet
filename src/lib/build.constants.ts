import Config from 'react-native-config'

const PRODUCT = {
  // define flags in alphabetical order
  FORCE_UPGRADE_BANNER: true,
  GRADIENT_TAB_BAR_ICON: true,
  IS_CHAT_CACHE_ENABLED: true,
  IS_SENTRY_ENABLED: false,
  IS_TIME_STATS_ENABLED: true,
  SHOW_ADMIN_RESET_UI: false,
  SHOW_BROADCAST_FILTER: false,
  SHOW_CHANNEL_UI: true,
  SHOW_CLEAN_DEVICE_SCREEN: false,
  SHOW_FULLSCREEN_AD: false,
  SHOW_MESSAGE_COPY_LINK: false,
  SHOW_NEW_PROFILE_FEATURE: false,
  SHOW_REMOVE_DEVICE: false,
  SHOW_RENAME_DEVICE: false,
  SHOW_PARTICIPANT_SHEET_USER_NAME: false,
  SHOW_PASSCODE: false,
  SHOW_PUSH_NOTIFICATION_OPTION: false,
  SHOW_ROOM_FILTER: false,
  SHOW_UPDATE_BANNER: true,
  SHOW_WALLET: false,
  TRANSPARENT_TAB_BAR: true,
}

// KEEP this eslint statement or will raise lint error when swith to other setup
const STAGING = {
  ...PRODUCT,
  // define flags in alphabetical order
  IS_CHAT_CACHE_ENABLED: true,
  IS_SENTRY_ENABLED: false,
  SHOW_BROADCAST_FILTER: false,
  SHOW_CHANNEL_UI: true,
  SHOW_CLEAN_DEVICE_SCREEN: true,
  SHOW_MESSAGE_COPY_LINK: true,
  SHOW_NEW_PROFILE_FEATURE: true,
  SHOW_PARTICIPANT_SHEET_USER_NAME: false,
  SHOW_PASSCODE: true,
  SHOW_PUSH_NOTIFICATION_OPTION: false,
  SHOW_ROOM_FILTER: true,
  SHOW_WALLET: true,
  SHOW_ADMIN_RESET_UI: true,
  SHOW_REMOVE_DEVICE: true,
  SHOW_RENAME_DEVICE: true,
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const SV = {
  ...PRODUCT,
  // define flags in alphabetical order
  GRADIENT_TAB_BAR_ICON: false,
  SHOW_PARTICIPANT_SHEET_USER_NAME: false,
  SHOW_PUSH_NOTIFICATION_OPTION: false,
  SHOW_ROOM_FILTER: true,
  SHOW_UPDATE_BANNER: false,
  TRANSPARENT_TAB_BAR: false,
  IS_CHAT_CACHE_ENABLED: true,
}

// Specify a feature set alphabetically
const APP = STAGING

export const FORCE_UPGRADE_BANNER = APP?.FORCE_UPGRADE_BANNER

export const GRADIENT_TAB_BAR_ICON = APP?.GRADIENT_TAB_BAR_ICON

export const IS_SENTRY_ENABLED = APP?.IS_SENTRY_ENABLED

export const SHOW_BROADCAST_FILTER = APP?.SHOW_BROADCAST_FILTER

export const SHOW_CHANNEL_UI = APP?.SHOW_CHANNEL_UI

export const SHOW_CLEAN_DEVICE_SCREEN = APP?.SHOW_CLEAN_DEVICE_SCREEN

export const SHOW_FULLSCREEN_AD = APP?.SHOW_FULLSCREEN_AD

export const SHOW_MESSAGE_COPY_LINK = APP?.SHOW_MESSAGE_COPY_LINK

export const SHOW_NEW_PROFILE_FEATURE = APP?.SHOW_NEW_PROFILE_FEATURE

export const SHOW_PARTICIPANT_SHEET_USER_NAME =
  APP?.SHOW_PARTICIPANT_SHEET_USER_NAME

export const SHOW_PASSCODE = APP?.SHOW_PASSCODE

export const SHOW_PUSH_NOTIFICATION_OPTION = APP?.SHOW_PUSH_NOTIFICATION_OPTION

export const SHOW_ROOM_FILTER = APP?.SHOW_ROOM_FILTER

export const SHOW_UPDATE_BANNER = APP?.SHOW_UPDATE_BANNER

export const SHOW_WALLET = APP?.SHOW_WALLET

export const TRANSPARENT_TAB_BAR = APP?.TRANSPARENT_TAB_BAR

export const IS_EXPERIMENTAL_BUILD = Config?.KEET_VARIANT === 'experimental'

export const IS_CHAT_CACHE_ENABLED = APP?.IS_CHAT_CACHE_ENABLED

export const IS_TIME_STATS_ENABLED = APP?.IS_TIME_STATS_ENABLED

export const SHOW_ADMIN_RESET_UI = APP?.SHOW_ADMIN_RESET_UI

export const SHOW_REMOVE_DEVICE = APP?.SHOW_REMOVE_DEVICE

export const SHOW_RENAME_DEVICE = APP?.SHOW_RENAME_DEVICE
