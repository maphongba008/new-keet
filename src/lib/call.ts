import { Alert } from 'react-native'

import { callLeave } from '@holepunchto/keet-store/store/call'

import type { Strings } from 'i18n/strings'

import { doVibrateSuccess } from './haptics'
import { getDispatch } from './store'

type showCallOnGoingAlertProps = {
  onPressLeave: Function
  onPressStay?: Function
  alertDesc?: string
  strings: Strings
}

export const showCallOnGoingAlert = ({
  onPressLeave,
  onPressStay,
  alertDesc,
  strings,
}: showCallOnGoingAlertProps) => {
  Alert.alert(
    strings.call.alertTitle,
    alertDesc || strings.call.alertDesc,
    [
      {
        text: strings.call.stayInCall,
        onPress: () => {
          onPressStay?.()
        },
        style: 'cancel',
      },
      {
        text: strings.call.leaveCall,
        onPress: () => {
          onPressLeave?.()
        },
        style: 'destructive',
      },
    ],
    { cancelable: true },
  )
}

export const dismissCall = () => {
  try {
    const dispatch = getDispatch()
    dispatch(callLeave())
    doVibrateSuccess()
  } catch (err) {
    console.log('Error in dismissCall', err)
  }
}
