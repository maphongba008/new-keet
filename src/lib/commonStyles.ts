import { Dimensions, I18nManager, StyleSheet } from 'react-native'

import { isAndroid } from './platform'

export const UI_SIZE_2 = 2
export const UI_SIZE_4 = 4
export const UI_SIZE_6 = 6
export const UI_SIZE_8 = 8
export const UI_SIZE_10 = 10
export const UI_SIZE_12 = 12
export const UI_SIZE_14 = 14
export const UI_SIZE_16 = 16
export const UI_SIZE_18 = 18
export const UI_SIZE_20 = 20
export const UI_SIZE_24 = 24
export const UI_SIZE_32 = 32
export const UI_SIZE_36 = 36
export const UI_SIZE_40 = 40
export const UI_SIZE_42 = 42
export const UI_SIZE_44 = 44
export const UI_SIZE_48 = 48
export const UI_SIZE_64 = 64
export const UI_SIZE_120 = 120

export const ICON_SIZE_8 = 8
export const ICON_SIZE_14 = 14
export const ICON_SIZE_15 = 15
export const ICON_SIZE_16 = 16
export const ICON_SIZE_20 = 20
export const ICON_SIZE_24 = 24
export const ICON_SIZE_28 = 28
export const ICON_SIZE_32 = 32
export const ICON_SIZE_44 = 44
export const ICON_SIZE_64 = 64
export const ICON_SIZE_72 = 72
export const ICON_SIZE_90 = 90
export const ICON_SIZE_112 = 112
export const ICON_SIZE_120 = 120
export const ICON_SIZE_130 = 130
export const ICON_SIZE_512 = 512

export const FONT_SIZE_100 = '100'
export const FONT_SIZE_200 = '200'
export const FONT_SIZE_300 = '300'
export const FONT_SIZE_400 = '400'
export const FONT_SIZE_500 = '500'
export const FONT_SIZE_600 = '600'
export const FONT_SIZE_700 = '700'
export const FONT_SIZE_800 = '800'
export const FONT_SIZE_900 = '900'
export const FONT_SIZE_BOLD = 'bold'
export const FONT_SIZE_BLACK = 'black'
export const FONT_SIZE_CONDENSED = 'condensed'

export const DEFAULT_AVATAR_COLORS = [
  '#7994F2',
  '#F2B379',
  '#F27983',
  '#2AAAAA',
]
export const NETWORK_WARNING = '#F2B414'
export const TRANSPARENT = 'transparent'
export const INPUT_ICON_COLOR = 'rgba(50, 58, 76, 0.5)'
export const ADMIN_WARNING_CONTAINER = 'rgba(255, 191, 0, 0.2)' // #FFBF00 with 20% transparency
export const INVITE_TEXT = '#646C7E'
export const BORDER_SEPARATOR_COLOR = '#1F2430'
export const OUTLINE_BUTTON_COLOR = '#FFA597'
export const SENDER_HIGHLIGHT_BG_COLOR = '#288DA8'
export const RECEIVER_HIGHLIGHT_BG_COLOR = '#3F424D'

export const { width, height } = Dimensions.get('window')

export const DIRECTION = I18nManager.isRTL ? 'right' : 'left'
export const DIRECTION_REVERSE = I18nManager.isRTL ? 'left' : 'right'
export const DIRECTION_CODE = I18nManager.isRTL ? 'rtl' : 'ltr'
export const LAYOUT_MARK = isAndroid ? (I18nManager.isRTL ? '‏' : '‎') : ''

export const commonStyles = StyleSheet.create({
  absolute: { position: 'absolute' },
  absoluteFill: StyleSheet.absoluteFillObject,
  alignItemsCenter: { alignItems: 'center' },
  alignItemsEnd: { alignItems: 'flex-end' },
  alignItemsStart: { alignItems: 'flex-start' },
  alignSelfCenter: { alignSelf: 'center' },
  alignSelfEnd: { alignSelf: 'flex-end' },
  alignSelfStart: { alignSelf: 'flex-start' },
  alignSelfStretch: { alignSelf: 'stretch' },
  bidirectionalInput: {
    textAlign: DIRECTION,
    writingDirection: DIRECTION_CODE,
  },
  capitalize: { textTransform: 'capitalize' },
  centerAlignedRow: {
    alignItems: 'center',
    flexDirection: 'row',
  },
  centeredLayout: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  column: { flexDirection: 'column' },
  container: { flex: 1 },
  // display: 'none' not always works correctly when component becomes visible again
  flex: { display: 'flex' },
  flex0: { flex: 0 },
  flex0_5: { flex: 0.5 },
  flexGrow: { flexGrow: 1 },
  flexSpaceAround: { justifyContent: 'space-around' },
  flexSpaceBetween: { justifyContent: 'space-between' },
  fullHeight: { height: '100%' },
  fullWidth: { width: '100%' },
  gap12: { gap: UI_SIZE_12 },
  hidden: { opacity: 0, position: 'absolute' },
  justifyCenter: {
    justifyContent: 'center',
  },
  justifyEnd: {
    justifyContent: 'flex-end',
  },
  justifyStart: {
    justifyContent: 'flex-start',
  },
  noMargin: { margin: 0 },
  noPadding: { padding: 0 },
  overflowHidden: { overflow: 'hidden' },
  overflowVisible: { overflow: 'visible' },
  posRelative: { position: 'relative' },
  row: { flexDirection: 'row' },
  rowCenterStart: {
    alignItems: 'flex-start',
    flexDirection: 'row',
    justifyContent: 'center',
  },
  rowReverse: { flexDirection: 'row-reverse' },
  rowStartCenter: {
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'flex-start',
  },
  textAlignCenter: { textAlign: 'center' },
  textAlignLeft: { textAlign: 'left' },
  textAlignRight: { textAlign: 'right' },
  textStrike: { textDecorationLine: 'line-through' },
  textUnderline: { textDecorationLine: 'underline' },
  uppercase: { textTransform: 'uppercase' },
  wrapFlex: { flexWrap: 'wrap' },
})

export default commonStyles
