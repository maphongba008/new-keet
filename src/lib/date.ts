import dayjs from 'dayjs'
import duration from 'dayjs/plugin/duration'
import isToday from 'dayjs/plugin/isToday'
import isYesterday from 'dayjs/plugin/isYesterday'
import LocalizedFormat from 'dayjs/plugin/localizedFormat'
import relativeTime from 'dayjs/plugin/relativeTime'
import utc from 'dayjs/plugin/utc'
import weekday from 'dayjs/plugin/weekday'

import {
  DATEJS_SUPPORT_LANGUAGES,
  DEFAULT_LANGUAGE,
  getStrings,
} from 'i18n/strings'

import { Keys, localStorage } from './localStorage'

import 'dayjs/locale/ar'
import 'dayjs/locale/de'
import 'dayjs/locale/es'
import 'dayjs/locale/fr'
import 'dayjs/locale/it'
import 'dayjs/locale/ka'
import 'dayjs/locale/nl'
import 'dayjs/locale/pt'
import 'dayjs/locale/ro'
import 'dayjs/locale/ru'
import 'dayjs/locale/tr'
import 'dayjs/locale/uk'
import 'dayjs/locale/vi'
import 'dayjs/locale/zh-cn'
import 'dayjs/locale/zh-tw'

const currentLocale =
  localStorage.getItem(Keys.I18N_LOCALE_KEY) ?? DEFAULT_LANGUAGE
if (currentLocale && DATEJS_SUPPORT_LANGUAGES[currentLocale]) {
  // Redundant checking (just putting here for future) because
  dayjs.locale(currentLocale) // so far datejs lang and our mobile app's lang are of exact match
}

dayjs.extend(relativeTime)
dayjs.extend(isToday)
dayjs.extend(isYesterday)
dayjs.extend(LocalizedFormat)
dayjs.extend(weekday)
dayjs.extend(duration)
dayjs.extend(utc)

export const timestampToRelativeTimeString = (date: number) => dayjs().to(date)

export const timestampToTimeString = (date: number) =>
  dayjs(date).format('HH:mm')

export const timestampToDay = (ts: number): string => {
  const date = dayjs(ts)
  const today = dayjs()
  const diff = today.diff(date, 'days')

  return date.isToday()
    ? getStrings().date.today
    : date.isYesterday()
      ? getStrings().date.yesterday
      : diff < 6
        ? date.format('dddd, D')
        : date.format('LL')
}

export const elapsedTime = (present: number, past: number): string => {
  const diff = present - past

  if (diff < 1000) {
    return `+${diff}ms`
  } else if (diff > 1000 && diff < 60000) {
    const secs =
      Math.round(dayjs(present).diff(past, 'second', true) * 100) / 100
    return `+${secs}s`
  }
  return `+${dayjs(present).diff(past, 'minutes')}m`
}

export const timeToLastMessage = (timestamp: number): string => {
  const {
    date: { short },
  } = getStrings()
  const date = dayjs(timestamp)
  const now = dayjs()

  const units: { unit: dayjs.UnitType; label: string }[] = [
    { unit: 'year', label: short.years },
    { unit: 'month', label: short.months },
    { unit: 'day', label: short.days },
    { unit: 'hour', label: short.hours },
    { unit: 'minute', label: short.minutes },
    { unit: 'second', label: short.seconds },
  ]

  for (const { unit, label } of units) {
    const diff = Math.max(0, now.diff(date, unit))
    if (diff) {
      return `${diff}${label}`
    }
  }

  return short.now
}

/**
 * Calculates the expiry date based on the specified duration in hours, days, and weeks,
 * and formats the result in "YYYY/M/D H:mm UTC".
 * Uses dayjs for precise and easy date manipulation and formatting.
 *
 * @param durationString string containing '2h'(2 hours) or '2d'(2 days) or '2w'(2 weeks) as strings. Each value represents
 *                        the duration that will be added to the current date to calculate the expiry date.
 * @returns The expiry date formatted as "YYYY/M/D H:mm UTC".
 */
export function calculateExpiryDateAndFormat(durationString: string): string {
  // Extract the numeric value and unit from the duration string
  const value = parseInt(durationString, 10)
  const unit = durationString.slice(-1) // Get the last character of the string
  // Calculate the duration based on the unit
  let durationFromUnit
  switch (unit) {
    case 'h':
      durationFromUnit = dayjs.duration(value, 'hour')
      break
    case 'd':
      durationFromUnit = dayjs.duration(value, 'day')
      break
    case 'w':
      durationFromUnit = dayjs.duration(value * 7, 'day') // Convert weeks to days
      break
    default:
      throw new Error('Invalid duration unit')
  }

  // Add the total duration to the current date and convert to UTC
  const expiryDate = dayjs().add(durationFromUnit).utc()

  // Format the date as "YYYY/M/D H:mm UTC".
  return expiryDate.format('YYYY/M/D H:mm') + ' UTC'
}

export const timestampToOutOfOrderTimestamp = (ts: number): string => {
  const date = dayjs(ts)

  return `${getStrings().chat.sent} ${date.format('dddd D')}`
}
