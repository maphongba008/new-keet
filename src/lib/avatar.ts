import { ViewStyle } from 'react-native'
import b4a from 'b4a'

import { DEFAULT_AVATAR_COLORS } from './commonStyles'

export const SEED_BITS = 32
export const EMOJI_REGEXP = /\p{Emoji}/u

export const name2seed = (name: string) => {
  if (name.length === SEED_BITS) {
    return name
  } else if (name.length > SEED_BITS) {
    return name.substring(0, SEED_BITS - 1)
  }
  return name.padEnd(SEED_BITS)
}

function randInt(min: number, max: number, seed?: string): number {
  const rand = seed
    ? b4a.readUInt32LE(b4a.from(seed, 'hex')) / (2 ** SEED_BITS - 1)
    : Math.random()
  min = Math.ceil(min)
  max = Math.floor(max)
  return Math.floor(rand * (max - min)) + min
}

export const randomColor = (seed?: string) => {
  return DEFAULT_AVATAR_COLORS[randInt(0, DEFAULT_AVATAR_COLORS.length, seed)]
}

export function getScaledFontSize({
  width,
  height,
}: Pick<ViewStyle, 'width' | 'height'>): number {
  const baseWidth = 30
  const baseHeight = 30
  const baseFontSize = 12
  const maxFontThreshold = 40

  if (typeof height !== 'number' || typeof width !== 'number') {
    return baseFontSize
  }

  // Calculate scaling factor
  const scalingFactor = (width * height) / (baseWidth * baseHeight)

  // Scale font size
  const fontSize = baseFontSize * Math.sqrt(scalingFactor)
  const minThreshold = Math.max(fontSize, baseFontSize)

  return Math.min(minThreshold, maxFontThreshold)
}
