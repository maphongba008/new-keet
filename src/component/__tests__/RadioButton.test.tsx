import { fireEvent, render, screen } from '@testing-library/react-native'

import LabeledRadio from 'component/RadioButton'
import { APPIUM_IDs } from 'lib/appium'

const mockOnPress = jest.fn()

const defaultProps = {
  value: true,
  onChange: mockOnPress,
}

describe('Radio button component tests', () => {
  it('should be checked when value is true', () => {
    render(<LabeledRadio value={true} onChange={mockOnPress} />)
    expect(
      screen.getByTestId(APPIUM_IDs.radio_btn_checked),
    ).not.toBeEmptyElement()
    expect(screen.toJSON()).toMatchSnapshot()
  })
  it('should be unchecked when value is false', () => {
    render(<LabeledRadio value={false} onChange={mockOnPress} />)
    expect(screen.getByTestId(APPIUM_IDs.radio_btn_checked)).toBeEmptyElement()
    expect(screen.toJSON()).toMatchSnapshot()
  })
  it('toggles correctly when clicked on', () => {
    render(<LabeledRadio value={false} onChange={mockOnPress} />)
    const button = screen.getByRole('button')
    fireEvent.press(button)
    expect(mockOnPress).toHaveBeenCalled()
  })
  it('renders Radio Button label correctly', () => {
    const btnLabel = 'Test Radio Label'
    render(<LabeledRadio label={btnLabel} {...defaultProps} />)
    expect(screen.getByText(btnLabel)).toBeOnTheScreen()
  })
})
