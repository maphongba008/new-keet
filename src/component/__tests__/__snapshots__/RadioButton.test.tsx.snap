// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio button component tests should be checked when value is true 1`] = `
<View
  accessibilityLabel="radio_btn"
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "opacity": 1,
      },
      {
        "alignItems": "center",
        "flexDirection": "row",
      },
    ]
  }
  testID="radio_btn"
>
  <View
    style={
      {
        "alignItems": "center",
        "borderColor": "#26D2E8",
        "borderRadius": 10,
        "borderWidth": 1,
        "height": 20,
        "justifyContent": "center",
        "marginRight": 8,
        "width": 20,
      }
    }
    testID="radio_btn_checked"
  >
    <View
      style={
        {
          "backgroundColor": "#26D2E8",
          "borderRadius": 9,
          "height": 9,
          "width": 9,
        }
      }
    />
  </View>
  <Text
    style={
      {
        "color": "#ffffff",
        "fontFamily": "OpenSans-Regular",
        "fontSize": 14,
        "includeFontPadding": false,
      }
    }
  />
</View>
`;

exports[`Radio button component tests should be unchecked when value is false 1`] = `
<View
  accessibilityLabel="radio_btn"
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "opacity": 1,
      },
      {
        "alignItems": "center",
        "flexDirection": "row",
      },
    ]
  }
  testID="radio_btn"
>
  <View
    style={
      {
        "alignItems": "center",
        "borderColor": "#26D2E8",
        "borderRadius": 10,
        "borderWidth": 1,
        "height": 20,
        "justifyContent": "center",
        "marginRight": 8,
        "width": 20,
      }
    }
    testID="radio_btn_checked"
  />
  <Text
    style={
      {
        "color": "#ffffff",
        "fontFamily": "OpenSans-Regular",
        "fontSize": 14,
        "includeFontPadding": false,
      }
    }
  />
</View>
`;
